rest-api/src/main/java/com/cloudbees/workflow/rest/endpoints/flownode/Describe.java

```diff
@@ -23,48 +23,51 @@
  */
 package com.cloudbees.workflow.rest.endpoints.flownode;
 
-import com.cloudbees.workflow.flownode.FlowNodeUtil;
 import com.cloudbees.workflow.rest.external.AtomFlowNodeExt;
+import com.cloudbees.workflow.rest.external.ChunkVisitor;
 import com.cloudbees.workflow.rest.external.FlowNodeExt;
 import com.cloudbees.workflow.rest.external.RunExt;
 import com.cloudbees.workflow.rest.external.StageNodeExt;
 import com.cloudbees.workflow.rest.endpoints.FlowNodeAPI;
-import com.google.common.cache.Cache;
-import com.google.common.cache.CacheBuilder;
+import hudson.model.Queue;
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
 import org.jenkinsci.plugins.workflow.graph.AtomNode;
 import org.jenkinsci.plugins.workflow.graph.FlowNode;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
 
-import javax.annotation.CheckForNull;
 import java.io.IOException;
-import java.util.concurrent.TimeUnit;
-import java.util.logging.Logger;
 
 /**
  * {@link FlowNode} "describe" endpoint.
  * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
  */
 public class Describe {
-    private static final Logger LOGGER = Logger.getLogger(Describe.class.getName());
 
     public static String getUrl(FlowNode node) {
         return FlowNodeAPI.getUrl(node) + "/describe";
     }
 
     public static FlowNodeExt get(FlowNode node) {
         if (StageNodeExt.isStageNode(node)) {
-            // TODO figure out a way to use the cache when we have the node+FlowExecution but not its parent run
-            /* RunExt cachedRunData = FlowNodeUtil.getCachedRun(node.getExecution());
-            if (cachedRunData != null) {
-                for (StageNodeExt stage : cachedRunData.getStages()) {
-                    if (stage.getId().equals(node.getId())) {
-                        return stage;
+            try {
+                // Digest the WorkflowRun to get the stages, using cache if possible.
+                // Future optimization for big runs: use the blackLists in the ForkScanner only digest until the
+                // Node before the stage
+                Queue.Executable exec = node.getExecution().getOwner().getExecutable();
+                if (exec instanceof WorkflowRun) {
+                    WorkflowRun run = (WorkflowRun)exec;
+                    RunExt runExt = RunExt.create(run);
+                    for (StageNodeExt st : runExt.getStages()) {
+                        if (st.getId().equals(node.getId())) {
+                            return st;
+                        }
                     }
                 }
-            }*/
-
-            StageNodeExt stageNodeExt = StageNodeExt.create(node);
-            stageNodeExt.addStageFlowNodes(node);
-            return stageNodeExt;
+            } catch (IOException ioe) {
+                throw new RuntimeException(ioe);
```

Comment:
@oleg-nenashev It is completely a dirty hack (the way we get at the WorkflowRun), but the only way to do it since pipeline does not provide a better API for it. 

Please take take a closer look at the IOException case here, and I think you'll see why this exception is completely safe to handle in such a fashion -- it's more of a formality handling it at all, since prechecks here should prevent it ever being thrown.